### PR TITLE
Initial attempt to address WW-5101

### PR DESCRIPTION
**Pull request by JCgH4164838Gh792C124B5**
Initial attempt to address WW-5101
- Add comments to methods related to ResourceBundle cache clearing.
- Correct clearTomcatCache() so that it acts as intended for Tomcat 7-10.
- Add clearResourceBundleClassloaderCaches() method that uses standard Java API calls to perform the cache clearing.
- Adjust clearMap() logic to only change accessiblity when required.

### CODE REVIEW

**Comment by JCgH4164838Gh792C124B5** [comment]
Hello Apache Struts Team.

This is an an initial attempt to address the reported issue.  Basic local testing indicates it is working, with no illegal reflective access warnings seen with Java 11.  People may decide to eventually remove the `clearTomcatCache()` method in the future, but for the moment this PR attempts to correct its behaviour so that it acts as the previous code and comments indicate it should.

Please let me know if the PR looks OK when you have time.

File: core/src/main/java/com/opensymphony/xwork2/util/AbstractLocalizedTextProvider.java [hunk 1]
```diff
@@ -308,36 +305,90 @@ protected void reloadBundles(Map<String, Object> context) {
...
+     * 
+     * It relies on the {@link #TOMCAT_RESOURCE_ENTRIES_FIELD} field name, base class name 
+     * {@link #TOMCAT_WEBAPP_CLASSLOADER_BASE}. and descendant class names {@link #TOMCAT_WEBAPP_CLASSLOADER},
+     * {@link #TOMCAT_PARALLEL_WEBAPP_CLASSLOADER}, to keep the values identified in the constants.
+     * It appears to be valid for Tomcat versions 7-10 so far, but could become invalid at any time in the future
+     * when the resource handling logic in Tomcat changes.
+     * 
+     * Note: With Java 9+, calling this method may result in "Illegal reflective access" warnings.  Be aware 
+     *       its logic may fail in a future version of Java that blocks the reflection calls needed for this method.
+     * {<code></code>
```
Comment (sepe81) [inline core/src/main/java/com/opensymphony/xwork2/util/AbstractLocalizedTextProvider.java:339]:
```suggestion
```

this line can probably be omitted

File: core/src/main/java/com/opensymphony/xwork2/util/AbstractLocalizedTextProvider.java [hunk 2]
```diff
@@ -308,36 +305,90 @@ protected void reloadBundles(Map<String, Object> context) {
...
     private void clearTomcatCache() {
         ClassLoader loader = getCurrentThreadContextClassLoader();
         // no need for compilation here.
         Class cl = loader.getClass();
+        Class superCl = cl.getSuperclass();
 
         try {
-            if ("org.apache.catalina.loader.WebappClassLoader".equals(cl.getName())) {
-                clearMap(cl, loader, TOMCAT_RESOURCE_ENTRIES_FIELD);
+            if ( (TOMCAT_WEBAPP_CLASSLOADER.equals(cl.getName()) || TOMCAT_PARALLEL_WEBAPP_CLASSLOADER.equals(cl.getName())) &&
```
Comment (sepe81) [inline core/src/main/java/com/opensymphony/xwork2/util/AbstractLocalizedTextProvider.java:348]:
```suggestion
            if ((TOMCAT_WEBAPP_CLASSLOADER.equals(cl.getName()) || TOMCAT_PARALLEL_WEBAPP_CLASSLOADER.equals(cl.getName())) &&
```

consistent formatting

File: core/src/main/java/com/opensymphony/xwork2/util/AbstractLocalizedTextProvider.java [hunk 3]
```diff
@@ -308,36 +305,90 @@ protected void reloadBundles(Map<String, Object> context) {
...
+                    (superCl != null && TOMCAT_WEBAPP_CLASSLOADER_BASE.equals(superCl.getName())) ) {
```
Comment (sepe81) [inline core/src/main/java/com/opensymphony/xwork2/util/AbstractLocalizedTextProvider.java:349]:
```suggestion
                    (superCl != null && TOMCAT_WEBAPP_CLASSLOADER_BASE.equals(superCl.getName()))) {
```

**Review by lukaszlenart — APPROVED** [review]
LGTM 👍

**Comment by lukaszlenart** [comment]
No complains, so let's merge!

**Pull request merged**